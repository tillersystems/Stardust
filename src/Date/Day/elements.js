import styled, { css } from 'styled-components';
import { lighten } from 'polished';

export const Container = styled.div`
  font-size: ${({ theme: { fonts } }) => fonts.size.medium};
  line-height: ${({ theme: { fonts } }) => `${fonts.size.mediumInt + 1.6}rem`}

  cursor: pointer;

${({ shadowed }) =>
  shadowed &&
  css`
    & > ${Content} {
      color: ${({ theme: { palette } }) => palette.gray};
    }
  `};

${({ disabled }) =>
  disabled &&
  css`
    cursor: not-allowed;

    & > ${Content} {
      color: ${({ theme: { palette } }) => palette.gray};
    }
  `};

  ${({ selected }) =>
    selected &&
    css`
      & > ${Content} {
        color: ${({ theme: { palette } }) => palette.white};
        background: ${({ theme: { palette } }) => palette.blue};
      }
    `};

  ${({ highlighted }) =>
    highlighted &&
    css`
      background: ${({ theme: { palette } }) => lighten(0.4, palette.blue)};
    `}

  ${({ starting }) =>
    starting &&
    css`
      border-top-left-radius: ${({ theme: { dimensions } }) => dimensions.radius};
      border-bottom-left-radius: ${({ theme: { dimensions } }) => dimensions.radius};
    `};

  ${({ ending }) =>
    ending &&
    css`
      border-top-right-radius: ${({ theme: { dimensions } }) => dimensions.radius};
      border-bottom-right-radius: ${({ theme: { dimensions } }) => dimensions.radius};
    `};
`;

export const Content = styled.div`
  width: 100%;
  height: 100%;

  ${({ starting }) =>
    starting &&
    css`
      border-top-left-radius: ${({ theme: { dimensions } }) => dimensions.radius};
      border-bottom-left-radius: ${({ theme: { dimensions } }) => dimensions.radius};
    `};

  ${({ ending }) =>
    ending &&
    css`
      border-top-right-radius: ${({ theme: { dimensions } }) => dimensions.radius};
      border-bottom-right-radius: ${({ theme: { dimensions } }) => dimensions.radius};
    `};
`;
