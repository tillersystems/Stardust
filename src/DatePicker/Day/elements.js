import styled, { css } from 'styled-components';
import { transparentize } from 'polished';

export const Container = styled.div`
  margin-top: 0.8rem;

  &:not(:nth-child(7n)) {
    padding-right: 0.8rem;
  }

  font-size: ${({ theme: { fonts } }) => fonts.size.medium};
  color: ${({ theme: { palette } }) => palette.darkBlue};

  ${({ isInPath }) =>
    isInPath &&
    css`
      background: ${({ theme: { palette } }) => transparentize(0.86, palette.primary.default)};
    `}

  ${({ isStartEdge }) =>
    isStartEdge &&
    css`
      border-top-left-radius: 50%;
      border-bottom-left-radius: 50%;
    `}

  ${({ isEndEdge }) =>
    isEndEdge &&
    css`
      border-top-right-radius: 50%;
      border-bottom-right-radius: 50%;

      &:not(:nth-child(7n)) {
        padding-right: 0;
        margin-right: 0.8rem;
      }
    `}

  cursor: pointer;
`;

export const Content = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  flex-basis: 2.8rem;

  width: 2.8rem;
  height: 2.8rem;
  max-width: 2.8rem;

  border-radius: 50%;

  user-select: none;

  ${({ disabled, shadowed, isSelected, isStartEdge, isEndEdge }) =>
    !disabled &&
    !shadowed &&
    !isSelected &&
    !isStartEdge &&
    !isEndEdge &&
    css`
      &:hover {
        background: ${({ theme: { palette } }) => transparentize(0.86, palette.primary.default)};
      }
    `};

  ${({ shadowed }) =>
    shadowed &&
    css`
      color: ${({ theme: { palette } }) => palette.mediumGrey};
    `};

  ${({ disabled }) =>
    disabled &&
    css`
      cursor: not-allowed;

      color: ${({ theme: { palette } }) => palette.lightGrey};
    `};

  ${({ isInPath, isStartEdge, isEndEdge }) =>
    isInPath &&
    !isStartEdge &&
    !isEndEdge &&
    css`
      background: ${({ theme: { palette } }) => transparentize(0.86, palette.primary.default)};
    `}

  ${({ isSelected, isStartEdge, isEndEdge }) =>
    (isSelected || isStartEdge || isEndEdge) &&
    css`
      color: ${({ theme: { palette } }) => palette.white};
      background: ${({ theme: { palette } }) => palette.primary.default};
    `};
`;
