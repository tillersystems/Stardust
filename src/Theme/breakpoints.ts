/**
 * Defines application breakpoints.
 */

const xs = 0;
const sm = 576;
const md = 768;
const lg = 992;
const xl = 1200;

export const breakpoints = {
  xs,
  sm,
  md,
  lg,
  xl,
};
