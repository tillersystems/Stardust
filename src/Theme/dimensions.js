/**
 * Defines the theming dimensions.
 */

const radius = 0.4;
const tiny = 0.3;
const small = 0.7;
const medium = 1.5;
const big = 3.1;
const huge = 4.5;

export const dimensions = {
  radius: `${radius}rem`,

  tiny: `${tiny}rem`,
  small: `${small}rem`,
  medium: `${medium}rem`,
  big: `${big}rem`,
  huge: `${huge}rem`,

  radiusInt: radius,

  tinyInt: tiny,
  smallInt: small,
  mediumInt: medium,
  bigInt: big,
  hugeInt: huge,
};
