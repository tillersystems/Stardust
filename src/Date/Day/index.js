import React from 'react';
import PropTypes from 'prop-types';

import { Container, Content } from './elements';

/**
 * A single Day component.
 *
 * @param {luxon.DateTime} date - The date to display.
 * @param {boolean} selected - Whether the day is selected or not.
 * @param {boolean} selectionStart - Whether this is a selection start or not.
 * @param {boolean} selectionEnd - Whether this is a selection end or not.
 * @param {boolean} highlighted - Whether the day is highlighted or not.
 * @param {boolean} highlightStart - Whether this is a highlight start or not.
 * @param {boolean} highlightEnd - Whether this is a highlight end or not.
 * @param {boolean} shadowed - Whether the day is shadowed or not.
 * @param {boolean} disabled - Whether the day is disabled or not.
 * @param {function} onClick - Handler called when the day is clicked.
 * @param {function} onOver - Handler called when the mouse is over the day.
 *
 * @return {jsx}
 */
const Day = ({
  date,
  selected,
  selectionStart,
  selectionEnd,
  highlighted,
  highlightStart,
  highlightEnd,
  shadowed,
  disabled,
  onClick,
  onOver,
}) => (
  <Container
    selected={selected}
    highlighted={highlighted}
    shadowed={shadowed}
    disabled={disabled}
    starting={highlightStart}
    ending={highlightEnd}
    onClick={onClick}
    onMouseOver={onOver}
    onFocus={() => {}}
  >
    <Content starting={selectionStart} ending={selectionEnd}>
      {date.toLocaleString({ day: '2-digit' })}
    </Content>
  </Container>
);

/** Display name. */
Day.displayName = 'Day';

const { object, bool, func } = PropTypes;
/** Validation prop types. */
Day.propTypes = {
  date: object.isRequired,
  selected: bool,
  selectionStart: bool,
  selectionEnd: bool,
  highlighted: bool,
  highlightStart: bool,
  highlightEnd: bool,
  shadowed: bool,
  disabled: bool,
  onClick: func,
  onOver: func,
};

/** Default props. */
Day.defaultProps = {
  selected: false,
  selectionStart: false,
  selectionEnd: false,
  highlighted: false,
  highlightStart: false,
  highlightEnd: false,
  shadowed: false,
  disabled: false,
  onClick: () => {},
  onOver: () => {},
};

export default Day;
