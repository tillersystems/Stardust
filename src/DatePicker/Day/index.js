import React from 'react';
import PropTypes from 'prop-types';

import { Container, Content } from './elements';

/**
 * A single Day component.
 *
 * @param {node} children -
 * @param {node} disabled -
 * @param {boolean} isSelected - Whether the day is selected or not.
 * @param {boolean} isStartEdge -
 * @param {boolean} isEndEdge -
 * @param {boolean} isInPath -
 * @param {boolean} shadowed - Whether the day is shadowed or not.
 * @param {function} onClick - Handler called when the day is clicked.
 *
 * @return {jsx}
 */
const Day = ({
  disabled,
  isSelected,
  shadowed,
  onClick,
  children,
  isInPath,
  isStartEdge,
  isEndEdge,
}) => (
  <Container onClick={onClick} isInPath={isInPath} isStartEdge={isStartEdge} isEndEdge={isEndEdge}>
    <Content
      disabled={disabled}
      isSelected={isSelected}
      isStartEdge={isStartEdge}
      isEndEdge={isEndEdge}
      shadowed={shadowed}
    >
      {children}
    </Content>
  </Container>
);

/**
 * PropTypes Validation.
 */
Day.displayName = 'Day';

/**
 * Default props.
 */
const { bool, func, node } = PropTypes;
Day.propTypes = {
  disabled: bool,
  isSelected: bool,
  isStartEdge: bool,
  isEndEdge: bool,
  isInPath: bool,
  shadowed: bool,
  onClick: func,
  children: node.isRequired,
};

/**
 * Default props.
 */
Day.defaultProps = {
  disabled: false,
  isSelected: false,
  isStartEdge: false,
  isEndEdge: false,
  isInPath: false,
  shadowed: false,
  onClick: () => {},
};

export default Day;
