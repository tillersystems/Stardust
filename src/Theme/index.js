/**
 * Theme exported to the provider
 * This is globally available in styled-components when interpolating a function like so:
 * ${(props) => props.theme}
 * Or using import { withTheme } from 'styled-components';
 */

import { breakpoints } from './breakpoints';
import { palette } from './palette';
import { dimensions } from './dimensions';
import { fonts } from './fonts';

const Theme = {
  breakpoints,
  palette,
  dimensions,
  fonts,
};

export default Theme;
