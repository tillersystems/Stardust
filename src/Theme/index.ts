import { DefaultTheme } from 'styled-components';

/**
 * Theme exported to the provider
 * This is globally available in styled-components when interpolating a function like so:
 * ${(props) => props.theme}
 * Or using import { withTheme } from 'styled-components';
 */

import { breakpoints } from './breakpoints';
import { palette } from './palette';
import { dimensions } from './dimensions';
import { fonts } from './fonts';

const Theme: DefaultTheme = {
  breakpoints,
  palette,
  dimensions,
  fonts,
};

export default Theme;
