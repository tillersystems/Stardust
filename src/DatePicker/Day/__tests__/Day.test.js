import React from 'react';
import { fireEvent } from '@testing-library/react';
import { transparentize } from 'polished';

import Day from '..';
import Theme from '../../../Theme';

const dayNumber = '10';

describe('Day', () => {
  test('should render without a problem', () => {
    const { container } = render(<Day>{dayNumber}</Day>);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render a disabled day', () => {
    const { getByText } = render(<Day disabled>{dayNumber}</Day>);

    const dayContentNode = getByText(dayNumber);

    expect(dayContentNode).toHaveStyleRule('cursor', 'not-allowed');
    expect(dayContentNode).toHaveStyleRule('color', Theme.palette.lightGrey);
    expect(dayContentNode).not.toHaveStyleRule('background', {
      modifier: ':hover',
    });
  });

  test('should render a selected day', () => {
    const { getByText } = render(<Day isSelected>{dayNumber}</Day>);

    const dayContentNode = getByText(dayNumber);

    expect(dayContentNode).toHaveStyleRule('color', Theme.palette.white);
    expect(dayContentNode).toHaveStyleRule('background', Theme.palette.primary.default);
    expect(dayContentNode).not.toHaveStyleRule('background', {
      modifier: ':hover',
    });
  });

  test('should render a shadowed day', () => {
    const { getByText } = render(<Day shadowed>{dayNumber}</Day>);

    const dayContentNode = getByText(dayNumber);

    expect(dayContentNode).toHaveStyleRule('color', Theme.palette.mediumGrey);
    expect(dayContentNode).not.toHaveStyleRule('background', {
      modifier: ':hover',
    });
  });

  test('should render as startEdge', () => {
    const expectedRadius = '50%';
    const { container, getByText } = render(<Day isStartEdge>{dayNumber}</Day>);

    const dayContainerNode = container.firstChild;
    const dayContentNode = getByText(dayNumber);

    expect(dayContainerNode).toHaveStyleRule('border-top-left-radius', expectedRadius);
    expect(dayContainerNode).toHaveStyleRule('border-bottom-left-radius', expectedRadius);

    expect(dayContentNode).toHaveStyleRule('color', Theme.palette.white);
    expect(dayContentNode).toHaveStyleRule('background', Theme.palette.primary.default);
    expect(dayContentNode).not.toHaveStyleRule(
      'background',
      transparentize(0.86, Theme.palette.primary.default),
    );
    expect(dayContentNode).not.toHaveStyleRule('background', {
      modifier: ':hover',
    });
  });

  test('should render as endEdge', () => {
    const expectedRadius = '50%';
    const { container, getByText } = render(<Day isEndEdge>{dayNumber}</Day>);

    const dayContainerNode = container.firstChild;
    const dayContentNode = getByText(dayNumber);

    expect(dayContainerNode).toHaveStyleRule('border-top-right-radius', expectedRadius);
    expect(dayContainerNode).toHaveStyleRule('border-bottom-right-radius', expectedRadius);

    expect(dayContentNode).toHaveStyleRule('color', Theme.palette.white);
    expect(dayContentNode).toHaveStyleRule('background', Theme.palette.primary.default);
    expect(dayContentNode).not.toHaveStyleRule(
      'background',
      transparentize(0.86, Theme.palette.primary.default),
    );
    expect(dayContentNode).not.toHaveStyleRule('background', {
      modifier: ':hover',
    });
  });

  test('should render with diffrent style if is in path', () => {
    const { container } = render(<Day isInPath>{dayNumber}</Day>);

    const dayContainerNode = container.firstChild;

    expect(dayContainerNode).toHaveStyleRule(
      'background',
      transparentize(0.86, Theme.palette.primary.default),
    );
  });

  test('should respond to a click handler', () => {
    const spy = jest.fn();
    const { container } = render(<Day onClick={spy}>{dayNumber}</Day>);

    const dayContainerNode = container.firstChild;

    // Click on root element
    fireEvent.click(dayContainerNode);

    expect(spy).toHaveBeenCalledTimes(1);
  });
});
