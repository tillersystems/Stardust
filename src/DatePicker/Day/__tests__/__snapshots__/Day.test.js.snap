// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Day should render without a problem 1`] = `
.c0 {
  margin-top: 0.8rem;
  font-size: 1.4rem;
  color: hsl(213,17%,20%);
  cursor: pointer;
}

.c0:not(:nth-child(7n)) {
  padding-right: 0.8rem;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-basis: 2.8rem;
  -ms-flex-preferred-size: 2.8rem;
  flex-basis: 2.8rem;
  width: 2.8rem;
  height: 2.8rem;
  max-width: 2.8rem;
  border-radius: 50%;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1:hover {
  background: rgba(30,146,204,0.14);
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    10
  </div>
</div>
`;
